File: lib/src/core/change_detection/differs/default_keyvalue_differ.dart
library angular2.src.core.change_detection.differs.default_keyvalue_differ;

import "package:angular2/src/facade/collection.dart"
    show MapWrapper, StringMapWrapper;
import "package:angular2/src/facade/lang.dart"
    show stringify, looseIdentical, isJsObject, isBlank;
import "package:angular2/src/facade/exceptions.dart" show BaseException;
import "../change_detector_ref.dart" show ChangeDetectorRef;
import "../differs/keyvalue_differs.dart"
    show KeyValueDiffer, KeyValueDifferFactory;

class DefaultKeyValueDifferFactory implements KeyValueDifferFactory {
  bool supports(dynamic obj) {
    return obj is Map || isJsObject(obj);
  }

  KeyValueDiffer create(ChangeDetectorRef cdRef) {
    return new DefaultKeyValueDiffer();
  }

  const DefaultKeyValueDifferFactory();
}

class DefaultKeyValueDiffer implements KeyValueDiffer {
  Map<dynamic, dynamic> _records = new Map();
  KVChangeRecord _mapHead = null;
  KVChangeRecord _previousMapHead = null;
  KVChangeRecord _changesHead = null;
  KVChangeRecord _changesTail = null;
  KVChangeRecord _additionsHead = null;
  KVChangeRecord _additionsTail = null;
  KVChangeRecord _removalsHead = null;
  KVChangeRecord _removalsTail = null;
  bool get isDirty {
    return !identical(this._additionsHead, null) ||
        !identical(this._changesHead, null) ||
        !identical(this._removalsHead, null);
  }

  forEachItem(Function fn) {
    KVChangeRecord record;
    for (record = this._mapHead;
        !identical(record, null);
        record = record._next) {
      fn(record);
    }
  }

  forEachPreviousItem(Function fn) {
    KVChangeRecord record;
    for (record = this._previousMapHead;
        !identical(record, null);
        record = record._nextPrevious) {
      fn(record);
    }
  }

  forEachChangedItem(Function fn) {
    KVChangeRecord record;
    for (record = this._changesHead;
        !identical(record, null);
        record = record._nextChanged) {
      fn(record);
    }
  }

  forEachAddedItem(Function fn) {
    KVChangeRecord record;
    for (record = this._additionsHead;
        !identical(record, null);
        record = record._nextAdded) {
      fn(record);
    }
  }

  forEachRemovedItem(Function fn) {
    KVChangeRecord record;
    for (record = this._removalsHead;
        !identical(record, null);
        record = record._nextRemoved) {
      fn(record);
    }
  }

  dynamic diff(Map<dynamic, dynamic> map) {
    if (isBlank(map)) map = MapWrapper.createFromPairs([]);
    if (!(map is Map || isJsObject(map))) {
      throw new BaseException('''Error trying to diff \'${ map}\'''');
    }
    if (this.check(map)) {
      return this;
    } else {
      return null;
    }
  }

  onDestroy() {}
  bool check(Map<dynamic, dynamic> map) {
    this._reset();
    var records = this._records;
    KVChangeRecord oldSeqRecord = this._mapHead;
    KVChangeRecord lastOldSeqRecord = null;
    KVChangeRecord lastNewSeqRecord = null;
    bool seqChanged = false;
    this._forEach(map, (value, key) {
      var newSeqRecord;
      if (!identical(oldSeqRecord, null) && identical(key, oldSeqRecord.key)) {
        newSeqRecord = oldSeqRecord;
        if (!looseIdentical(value, oldSeqRecord.currentValue)) {
          oldSeqRecord.previousValue = oldSeqRecord.currentValue;
          oldSeqRecord.currentValue = value;
          this._addToChanges(oldSeqRecord);
        }
      } else {
        seqChanged = true;
        if (!identical(oldSeqRecord, null)) {
          oldSeqRecord._next = null;
          this._removeFromSeq(lastOldSeqRecord, oldSeqRecord);
          this._addToRemovals(oldSeqRecord);
        }
        if (records.containsKey(key)) {
          newSeqRecord = records[key];
        } else {
          newSeqRecord = new KVChangeRecord(key);
          records[key] = newSeqRecord;
          newSeqRecord.currentValue = value;
          this._addToAdditions(newSeqRecord);
        }
      }
      if (seqChanged) {
        if (this._isInRemovals(newSeqRecord)) {
          this._removeFromRemovals(newSeqRecord);
        }
        if (lastNewSeqRecord == null) {
          this._mapHead = newSeqRecord;
        } else {
          lastNewSeqRecord._next = newSeqRecord;
        }
      }
      lastOldSeqRecord = oldSeqRecord;
      lastNewSeqRecord = newSeqRecord;
      oldSeqRecord = identical(oldSeqRecord, null) ? null : oldSeqRecord._next;
    });
    this._truncate(lastOldSeqRecord, oldSeqRecord);
    return this.isDirty;
  }

  /** @internal */
  _reset() {
    if (this.isDirty) {
      KVChangeRecord record;
      // Record the state of the mapping
      for (record = this._previousMapHead = this._mapHead;
          !identical(record, null);
          record = record._next) {
        record._nextPrevious = record._next;
      }
      for (record = this._changesHead;
          !identical(record, null);
          record = record._nextChanged) {
        record.previousValue = record.currentValue;
      }
      for (record = this._additionsHead;
          record != null;
          record = record._nextAdded) {
        record.previousValue = record.currentValue;
      }
      // todo(vicb) once assert is supported

      // assert(() {

      //  var r = _changesHead;

      //  while (r != null) {

      //    var nextRecord = r._nextChanged;

      //    r._nextChanged = null;

      //    r = nextRecord;

      //  }

      //

      //  r = _additionsHead;

      //  while (r != null) {

      //    var nextRecord = r._nextAdded;

      //    r._nextAdded = null;

      //    r = nextRecord;

      //  }

      //

      //  r = _removalsHead;

      //  while (r != null) {

      //    var nextRecord = r._nextRemoved;

      //    r._nextRemoved = null;

      //    r = nextRecord;

      //  }

      //

      //  return true;

      //});
      this._changesHead = this._changesTail = null;
      this._additionsHead = this._additionsTail = null;
      this._removalsHead = this._removalsTail = null;
    }
  }

  /** @internal */
  _truncate(KVChangeRecord lastRecord, KVChangeRecord record) {
    while (!identical(record, null)) {
      if (identical(lastRecord, null)) {
        this._mapHead = null;
      } else {
        lastRecord._next = null;
      }
      var nextRecord = record._next;
      // todo(vicb) assert

      // assert((() {

      //  record._next = null;

      //  return true;

      //}));
      this._addToRemovals(record);
      lastRecord = record;
      record = nextRecord;
    }
    for (KVChangeRecord rec = this._removalsHead;
        !identical(rec, null);
        rec = rec._nextRemoved) {
      rec.previousValue = rec.currentValue;
      rec.currentValue = null;
      (this._records.containsKey(rec.key) &&
          (this._records.remove(rec.key) != null || true));
    }
  }

  /** @internal */
  _isInRemovals(KVChangeRecord record) {
    return identical(record, this._removalsHead) ||
        !identical(record._nextRemoved, null) ||
        !identical(record._prevRemoved, null);
  }

  /** @internal */
  _addToRemovals(KVChangeRecord record) {
    // todo(vicb) assert

    // assert(record._next == null);

    // assert(record._nextAdded == null);

    // assert(record._nextChanged == null);

    // assert(record._nextRemoved == null);

    // assert(record._prevRemoved == null);
    if (identical(this._removalsHead, null)) {
      this._removalsHead = this._removalsTail = record;
    } else {
      this._removalsTail._nextRemoved = record;
      record._prevRemoved = this._removalsTail;
      this._removalsTail = record;
    }
  }

  /** @internal */
  _removeFromSeq(KVChangeRecord prev, KVChangeRecord record) {
    var next = record._next;
    if (identical(prev, null)) {
      this._mapHead = next;
    } else {
      prev._next = next;
    }
  }

  /** @internal */
  _removeFromRemovals(KVChangeRecord record) {
    // todo(vicb) assert

    // assert(record._next == null);

    // assert(record._nextAdded == null);

    // assert(record._nextChanged == null);
    var prev = record._prevRemoved;
    var next = record._nextRemoved;
    if (identical(prev, null)) {
      this._removalsHead = next;
    } else {
      prev._nextRemoved = next;
    }
    if (identical(next, null)) {
      this._removalsTail = prev;
    } else {
      next._prevRemoved = prev;
    }
    record._prevRemoved = record._nextRemoved = null;
  }

  /** @internal */
  _addToAdditions(KVChangeRecord record) {
    // todo(vicb): assert

    // assert(record._next == null);

    // assert(record._nextAdded == null);

    // assert(record._nextChanged == null);

    // assert(record._nextRemoved == null);

    // assert(record._prevRemoved == null);
    if (identical(this._additionsHead, null)) {
      this._additionsHead = this._additionsTail = record;
    } else {
      this._additionsTail._nextAdded = record;
      this._additionsTail = record;
    }
  }

  /** @internal */
  _addToChanges(KVChangeRecord record) {
    // todo(vicb) assert

    // assert(record._nextAdded == null);

    // assert(record._nextChanged == null);

    // assert(record._nextRemoved == null);

    // assert(record._prevRemoved == null);
    if (identical(this._changesHead, null)) {
      this._changesHead = this._changesTail = record;
    } else {
      this._changesTail._nextChanged = record;
      this._changesTail = record;
    }
  }

  String toString() {
    var items = [];
    var previous = [];
    var changes = [];
    var additions = [];
    var removals = [];
    KVChangeRecord record;
    for (record = this._mapHead;
        !identical(record, null);
        record = record._next) {
      items.add(stringify(record));
    }
    for (record = this._previousMapHead;
        !identical(record, null);
        record = record._nextPrevious) {
      previous.add(stringify(record));
    }
    for (record = this._changesHead;
        !identical(record, null);
        record = record._nextChanged) {
      changes.add(stringify(record));
    }
    for (record = this._additionsHead;
        !identical(record, null);
        record = record._nextAdded) {
      additions.add(stringify(record));
    }
    for (record = this._removalsHead;
        !identical(record, null);
        record = record._nextRemoved) {
      removals.add(stringify(record));
    }
    return "map: " +
        items.join(", ") +
        "\n" +
        "previous: " +
        previous.join(", ") +
        "\n" +
        "additions: " +
        additions.join(", ") +
        "\n" +
        "changes: " +
        changes.join(", ") +
        "\n" +
        "removals: " +
        removals.join(", ") +
        "\n";
  }

  /** @internal */
  _forEach(obj, Function fn) {
    if (obj is Map) {
      ((obj as Map<dynamic, dynamic>))
          .forEach((k, v) => ((fn as dynamic))(v, k));
    } else {
      StringMapWrapper.forEach(obj, fn);
    }
  }
}

class KVChangeRecord {
  dynamic key;
  dynamic previousValue = null;
  dynamic currentValue = null;
  /** @internal */
  KVChangeRecord _nextPrevious = null;
  /** @internal */
  KVChangeRecord _next = null;
  /** @internal */
  KVChangeRecord _nextAdded = null;
  /** @internal */
  KVChangeRecord _nextRemoved = null;
  /** @internal */
  KVChangeRecord _prevRemoved = null;
  /** @internal */
  KVChangeRecord _nextChanged = null;
  KVChangeRecord(this.key) {}
  String toString() {
    return looseIdentical(this.previousValue, this.currentValue)
        ? stringify(this.key)
        : (stringify(this.key) +
            "[" +
            stringify(this.previousValue) +
            "->" +
            stringify(this.currentValue) +
            "]");
  }
}

File: modules/@angular/core/src/change_detection/differs/default_keyvalue_differ.ts
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */

import {StringMapWrapper} from '../../facade/collection';
import {isJsObject, looseIdentical, stringify} from '../../facade/lang';
import {ChangeDetectorRef} from '../change_detector_ref';

import {KeyValueChangeRecord, KeyValueChanges, KeyValueDiffer, KeyValueDifferFactory} from './keyvalue_differs';


export class DefaultKeyValueDifferFactory<K, V> implements KeyValueDifferFactory {
  constructor() {}
  supports(obj: any): boolean { return obj instanceof Map || isJsObject(obj); }

  create<K, V>(cdRef: ChangeDetectorRef): KeyValueDiffer<K, V> {
    return new DefaultKeyValueDiffer<K, V>();
  }
}

export class DefaultKeyValueDiffer<K, V> implements KeyValueDiffer<K, V>, KeyValueChanges<K, V> {
  private _records: Map<K, V> = new Map<K, V>();
  private _mapHead: KeyValueChangeRecord_<K, V> = null;
  private _previousMapHead: KeyValueChangeRecord_<K, V> = null;
  private _changesHead: KeyValueChangeRecord_<K, V> = null;
  private _changesTail: KeyValueChangeRecord_<K, V> = null;
  private _additionsHead: KeyValueChangeRecord_<K, V> = null;
  private _additionsTail: KeyValueChangeRecord_<K, V> = null;
  private _removalsHead: KeyValueChangeRecord_<K, V> = null;
  private _removalsTail: KeyValueChangeRecord_<K, V> = null;

  get isDirty(): boolean {
    return this._additionsHead !== null || this._changesHead !== null ||
        this._removalsHead !== null;
  }

  forEachItem(fn: (r: KeyValueChangeRecord<K, V>) => void) {
    let record: KeyValueChangeRecord_<K, V>;
    for (record = this._mapHead; record !== null; record = record._next) {
      fn(record);
    }
  }

  forEachPreviousItem(fn: (r: KeyValueChangeRecord<K, V>) => void) {
    let record: KeyValueChangeRecord_<K, V>;
    for (record = this._previousMapHead; record !== null; record = record._nextPrevious) {
      fn(record);
    }
  }

  forEachChangedItem(fn: (r: KeyValueChangeRecord<K, V>) => void) {
    let record: KeyValueChangeRecord_<K, V>;
    for (record = this._changesHead; record !== null; record = record._nextChanged) {
      fn(record);
    }
  }

  forEachAddedItem(fn: (r: KeyValueChangeRecord<K, V>) => void) {
    let record: KeyValueChangeRecord_<K, V>;
    for (record = this._additionsHead; record !== null; record = record._nextAdded) {
      fn(record);
    }
  }

  forEachRemovedItem(fn: (r: KeyValueChangeRecord<K, V>) => void) {
    let record: KeyValueChangeRecord_<K, V>;
    for (record = this._removalsHead; record !== null; record = record._nextRemoved) {
      fn(record);
    }
  }

  diff(map: Map<any, any>|{[k: string]: any}): any {
    if (!map) {
      map = new Map();
    } else if (!(map instanceof Map || isJsObject(map))) {
      throw new Error(`Error trying to diff '${map}'`);
    }

    return this.check(map) ? this : null;
  }

  onDestroy() {}

  check(map: Map<any, any>|{[k: string]: any}): boolean {
    this._reset();
    const records = this._records;
    let oldSeqRecord: KeyValueChangeRecord_<K, V> = this._mapHead;
    let lastOldSeqRecord: KeyValueChangeRecord_<K, V> = null;
    let lastNewSeqRecord: KeyValueChangeRecord_<K, V> = null;
    let seqChanged: boolean = false;

    this._forEach(map, (value: any, key: any) => {
      let newSeqRecord: any;
      if (oldSeqRecord && key === oldSeqRecord.key) {
        newSeqRecord = oldSeqRecord;
        this._maybeAddToChanges(newSeqRecord, value);
      } else {
        seqChanged = true;
        if (oldSeqRecord !== null) {
          this._removeFromSeq(lastOldSeqRecord, oldSeqRecord);
          this._addToRemovals(oldSeqRecord);
        }
        if (records.has(key)) {
          newSeqRecord = records.get(key);
          this._maybeAddToChanges(newSeqRecord, value);
        } else {
          newSeqRecord = new KeyValueChangeRecord_<K, V>(key);
          records.set(key, newSeqRecord);
          newSeqRecord.currentValue = value;
          this._addToAdditions(newSeqRecord);
        }
      }

      if (seqChanged) {
        if (this._isInRemovals(newSeqRecord)) {
          this._removeFromRemovals(newSeqRecord);
        }
        if (lastNewSeqRecord == null) {
          this._mapHead = newSeqRecord;
        } else {
          lastNewSeqRecord._next = newSeqRecord;
        }
      }
      lastOldSeqRecord = oldSeqRecord;
      lastNewSeqRecord = newSeqRecord;
      oldSeqRecord = oldSeqRecord && oldSeqRecord._next;
    });
    this._truncate(lastOldSeqRecord, oldSeqRecord);
    return this.isDirty;
  }

  /** @internal */
  _reset() {
    if (this.isDirty) {
      let record: KeyValueChangeRecord_<K, V>;
      // Record the state of the mapping
      for (record = this._previousMapHead = this._mapHead; record !== null; record = record._next) {
        record._nextPrevious = record._next;
      }

      for (record = this._changesHead; record !== null; record = record._nextChanged) {
        record.previousValue = record.currentValue;
      }

      for (record = this._additionsHead; record != null; record = record._nextAdded) {
        record.previousValue = record.currentValue;
      }

      this._changesHead = this._changesTail = null;
      this._additionsHead = this._additionsTail = null;
      this._removalsHead = this._removalsTail = null;
    }
  }

  private _truncate(lastRecord: KeyValueChangeRecord_<K, V>, record: KeyValueChangeRecord_<K, V>) {
    while (record !== null) {
      if (lastRecord === null) {
        this._mapHead = null;
      } else {
        lastRecord._next = null;
      }
      const nextRecord = record._next;
      this._addToRemovals(record);
      lastRecord = record;
      record = nextRecord;
    }

    for (let rec: KeyValueChangeRecord_<K, V> = this._removalsHead; rec !== null;
         rec = rec._nextRemoved) {
      rec.previousValue = rec.currentValue;
      rec.currentValue = null;
      this._records.delete(rec.key);
    }
  }

  private _maybeAddToChanges(record: KeyValueChangeRecord_<K, V>, newValue: any): void {
    if (!looseIdentical(newValue, record.currentValue)) {
      record.previousValue = record.currentValue;
      record.currentValue = newValue;
      this._addToChanges(record);
    }
  }

  private _isInRemovals(record: KeyValueChangeRecord_<K, V>) {
    return record === this._removalsHead || record._nextRemoved !== null ||
        record._prevRemoved !== null;
  }

  private _addToRemovals(record: KeyValueChangeRecord_<K, V>) {
    if (this._removalsHead === null) {
      this._removalsHead = this._removalsTail = record;
    } else {
      this._removalsTail._nextRemoved = record;
      record._prevRemoved = this._removalsTail;
      this._removalsTail = record;
    }
  }

  private _removeFromSeq(prev: KeyValueChangeRecord_<K, V>, record: KeyValueChangeRecord_<K, V>) {
    const next = record._next;
    if (prev === null) {
      this._mapHead = next;
    } else {
      prev._next = next;
    }
    record._next = null;
  }

  private _removeFromRemovals(record: KeyValueChangeRecord_<K, V>) {
    const prev = record._prevRemoved;
    const next = record._nextRemoved;
    if (prev === null) {
      this._removalsHead = next;
    } else {
      prev._nextRemoved = next;
    }
    if (next === null) {
      this._removalsTail = prev;
    } else {
      next._prevRemoved = prev;
    }
    record._prevRemoved = record._nextRemoved = null;
  }

  private _addToAdditions(record: KeyValueChangeRecord_<K, V>) {
    if (this._additionsHead === null) {
      this._additionsHead = this._additionsTail = record;
    } else {
      this._additionsTail._nextAdded = record;
      this._additionsTail = record;
    }
  }

  private _addToChanges(record: KeyValueChangeRecord_<K, V>) {
    if (this._changesHead === null) {
      this._changesHead = this._changesTail = record;
    } else {
      this._changesTail._nextChanged = record;
      this._changesTail = record;
    }
  }

  toString(): string {
    const items: any[] = [];
    const previous: any[] = [];
    const changes: any[] = [];
    const additions: any[] = [];
    const removals: any[] = [];
    let record: KeyValueChangeRecord_<K, V>;

    for (record = this._mapHead; record !== null; record = record._next) {
      items.push(stringify(record));
    }
    for (record = this._previousMapHead; record !== null; record = record._nextPrevious) {
      previous.push(stringify(record));
    }
    for (record = this._changesHead; record !== null; record = record._nextChanged) {
      changes.push(stringify(record));
    }
    for (record = this._additionsHead; record !== null; record = record._nextAdded) {
      additions.push(stringify(record));
    }
    for (record = this._removalsHead; record !== null; record = record._nextRemoved) {
      removals.push(stringify(record));
    }

    return 'map: ' + items.join(', ') + '\n' +
        'previous: ' + previous.join(', ') + '\n' +
        'additions: ' + additions.join(', ') + '\n' +
        'changes: ' + changes.join(', ') + '\n' +
        'removals: ' + removals.join(', ') + '\n';
  }

  /** @internal */
  private _forEach<K, V>(obj: Map<K, V>|{[k: string]: V}, fn: (v: V, k: any) => void) {
    if (obj instanceof Map) {
      obj.forEach(fn);
    } else {
      Object.keys(obj).forEach(k => fn(obj[k], k));
    }
  }
}


/**
 * @stable
 */
class KeyValueChangeRecord_<K, V> implements KeyValueChangeRecord<K, V> {
  previousValue: V = null;
  currentValue: V = null;

  /** @internal */
  _nextPrevious: KeyValueChangeRecord_<K, V> = null;
  /** @internal */
  _next: KeyValueChangeRecord_<K, V> = null;
  /** @internal */
  _nextAdded: KeyValueChangeRecord_<K, V> = null;
  /** @internal */
  _nextRemoved: KeyValueChangeRecord_<K, V> = null;
  /** @internal */
  _prevRemoved: KeyValueChangeRecord_<K, V> = null;
  /** @internal */
  _nextChanged: KeyValueChangeRecord_<K, V> = null;

  constructor(public key: K) {}

  toString(): string {
    return looseIdentical(this.previousValue, this.currentValue) ?
        stringify(this.key) :
        (stringify(this.key) + '[' + stringify(this.previousValue) + '->' +
         stringify(this.currentValue) + ']');
  }
}
